server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/NettyRequestBody.scala

```diff
@@ -1,52 +1,82 @@
-    monadError: MonadError[F]
-) extends RequestBody[F, NoStreams] {
+  implicit def monad: MonadError[F]
 
-  override val streams: capabilities.Streams[NoStreams] = NoStreams
+  /** Backend-specific implementation for creating a file. */
+  def createFile: ServerRequest => F[TapirFile]
 
-  override def toRaw[RAW](serverRequest: ServerRequest, bodyType: RawBodyType[RAW]): F[RawValue[RAW]] = {
+  /** Backend-specific way to process all elements emitted by a Publisher[HttpContent] into a raw array of bytes.
+    *
+    * @param publisher
+    *   reactive publisher emitting byte chunks.
+    * @param maxBytes
+    *   optional request length limit. If exceeded, The effect `F` is failed with a [[sttp.capabilities.StreamMaxLengthExceededException]]
+    * @return
+    *   An effect which finishes with a single array of all collected bytes.
+    */
+  def publisherToBytes(publisher: Publisher[HttpContent], maxBytes: Option[Long]): F[Array[Byte]]
 
-    /** [[ByteBufUtil.getBytes(io.netty.buffer.ByteBuf)]] copies buffer without affecting reader index of the original. */
-    def requestContentAsByteArray = ByteBufUtil.getBytes(nettyRequest(serverRequest).content())
+  /** Backend-specific way to process all elements emitted by a Publisher[HttpContent] and write their bytes into a file.
+    *
+    * @param serverRequest
+    *   can have underlying `Publisher[HttpContent]` or an empty `FullHttpRequest`
+    * @param file
+    *   an empty file where bytes should be stored.
+    * @param maxBytes
+    *   optional request length limit. If exceeded, The effect `F` is failed with a [[sttp.capabilities.StreamMaxLengthExceededException]]
+    * @return
+    *   an effect which finishes when all data is written to the file.
+    */
+  def writeToFile(serverRequest: ServerRequest, file: TapirFile, maxBytes: Option[Long]): F[Unit]
 
+  override def toRaw[RAW](serverRequest: ServerRequest, bodyType: RawBodyType[RAW], maxBytes: Option[Long]): F[RawValue[RAW]] = {
     bodyType match {
-      case RawBodyType.StringBody(charset) => monadError.unit(RawValue(nettyRequest(serverRequest).content().toString(charset)))
-      case RawBodyType.ByteArrayBody       => monadError.unit(RawValue(requestContentAsByteArray))
-      case RawBodyType.ByteBufferBody      => monadError.unit(RawValue(ByteBuffer.wrap(requestContentAsByteArray)))
-      case RawBodyType.InputStreamBody     => monadError.unit(RawValue(new ByteBufInputStream(nettyRequest(serverRequest).content())))
+      case RawBodyType.StringBody(charset) => readAllBytes(serverRequest, maxBytes).map(bs => RawValue(new String(bs, charset)))
+      case RawBodyType.ByteArrayBody =>
+        readAllBytes(serverRequest, maxBytes).map(RawValue(_))
+      case RawBodyType.ByteBufferBody =>
+        readAllBytes(serverRequest, maxBytes).map(bs => RawValue(ByteBuffer.wrap(bs)))
+      case RawBodyType.InputStreamBody =>
+        // Possibly can be optimized to avoid loading all data eagerly into memory
```

Comment:
let's add a follow-up task for that, it might be quite surprising that this behaves this way

Reply:
https://github.com/softwaremill/tapir/issues/3366